File: money-calculator-back/src/main/java/com/moneycalculator/back/models/Transaction.java
package com.moneycalculator.back.models;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;


@Data
@Entity
@AllArgsConstructor
@NoArgsConstructor
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "type", discriminatorType = DiscriminatorType.STRING)
@Table(name = "transactions")
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "_id")
    private Integer id;

    private String label;

    private Double amount;

    @Column(name = "type", insertable = false, updatable = false)
    private String type;

    @ManyToOne
    @JoinColumn(name = "account_id", referencedColumnName = "_id", nullable = false)
    private Account account;
}

File: money-calculator-front/src/models/Transaction.ts
import {Account} from "@/models/Account";

export enum TransactionType {
    EXPENSE = "expense",
    INCOME = "income",
}

export interface Transaction {
    id: number;
    label: string;
    amount: number;
    account: Account;
    type: TransactionType;
}

export interface Income extends Transaction {
    type: TransactionType.INCOME;
}

export interface Expense extends Transaction {
    type: TransactionType.EXPENSE;
}
